File: yt/yt/flow/examples/kotlin/static_table_join/static_table_join/src/main/kotlin/tech/ytsaurus/flow/examples/statictablejoin/Enricher.kt
package tech.ytsaurus.flow.examples.statictablejoin

import tech.ytsaurus.flow.computation.OutputCollector
import tech.ytsaurus.flow.context.RuntimeContext
import tech.ytsaurus.flow.function.RowFunction
import tech.ytsaurus.flow.row.ExtendedMessage
import tech.ytsaurus.flow.state.JoinedExternalStateDescriptor
import tech.ytsaurus.flow.state.StateDescriptors

/**
 * Joins each event against the reference state and emits the enriched row.
 */
class Enricher : RowFunction {
    companion object {
        private val REFERENCE_STATE: JoinedExternalStateDescriptor =
            StateDescriptors.externalReadOnly("/reference_state")
    }

    // [BEGIN on_message]
    override fun onMessage(message: ExtendedMessage, output: OutputCollector, ctx: RuntimeContext) {
        val stateAccessor = ctx.getState(REFERENCE_STATE, message)
        val normalizedName = stateAccessor.orDefault.get("normalized_name", String::class.java) ?: return

        output.addMessage(
            ctx.createMessageBuilder("enriched")
                .set("key", message.get("key", Long::class.java))
                .set("name", normalizedName)
                .finish()
        )
    }
    // [END on_message]
}

File: yt/yt/flow/examples/go/static_table_join/enricher.go
package main

import (
	"context"
	"errors"

	"go.ytsaurus.tech/yt/go/flow"
)

const enrichedStreamID = "enriched"

// [BEGIN enricher]

type enricher struct{}

var _ flow.RowFunction = (*enricher)(nil)

func (*enricher) OnMessage(
	ctx context.Context,
	rt flow.Runtime,
	msg flow.ExtendedMessage,
	out flow.OutputCollector,
) error {
	var event eventMessage
	if err := msg.ConvertTo(&event); err != nil {
		return err
	}

	name, joined, err := joinedName(rt, msg)
	if err != nil || !joined {
		return err
	}

	enriched := flow.NewYSONMessage[enrichedMessage](enrichedStreamID)
	enriched.Key = event.Key
	enriched.Name = name
	encoded, err := flow.ConvertFrom(rt, enriched)
	if err != nil {
		return err
	}
	out.AddMessage(encoded)
	return nil
}

func joinedName(rt flow.Runtime, msg flow.ExtendedMessage) (string, bool, error) {
	state, err := flow.OpenJoinedExternalState(rt, referenceStateName, msg)
	if errors.Is(err, flow.ErrStateNotRead) {
		return "", false, nil
	}
	if err != nil {
		return "", false, err
	}

	var reference referenceState
	exists, err := state.ConvertTo(&reference)
	if err != nil || !exists || reference.NormalizedName == nil {
		return "", false, err
	}
	return *reference.NormalizedName, true, nil
}

// [END enricher]
